pkg/actions/create/nodegroup.go

```diff
@@ -0,0 +1,307 @@
+			logger.Debug("cluster has withOIDC enabled but is not using IRSA for CNI, will add CNI policy to node role")
+		}
+
+		nodeGroupTasks := stackManager.NewUnmanagedNodeGroupTask(cfg.NodeGroups, supportsManagedNodes, !awsNodeUsesIRSA)
+		if nodeGroupTasks.Len() > 0 {
+			allNodeGroupTasks.Append(nodeGroupTasks)
+		}
+		managedTasks := stackManager.NewManagedNodeGroupTask(cfg.ManagedNodeGroups, !awsNodeUsesIRSA)
+		if managedTasks.Len() > 0 {
+			allNodeGroupTasks.Append(managedTasks)
+		}
+
+		tasks.Append(allNodeGroupTasks)
+		logger.Info(tasks.Describe())
+		errs := tasks.DoAllSync()
+		if len(errs) > 0 {
+			logger.Info("%d error(s) occurred and nodegroups haven't been created properly, you may wish to check CloudFormation console", len(errs))
+			logger.Info("to cleanup resources, run 'eksctl delete nodegroup --region=%s --cluster=%s --name=<name>' for each of the failed nodegroup", cfg.Metadata.Region, cfg.Metadata.Name)
+			for _, err := range errs {
+				if err != nil {
+					logger.Critical("%s\n", err.Error())
+				}
+			}
+			return fmt.Errorf("failed to create nodegroups for cluster %q", cfg.Metadata.Name)
+		}
+	}
+
+	{ // post-creation action
+		tasks := ctl.ClusterTasksForNodeGroups(cfg, c.options.InstallNeuronDevicePlugin)
+		logger.Info(tasks.Describe())
+		errs := tasks.DoAllSync()
+		if len(errs) > 0 {
+			logger.Info("%d error(s) occurred and nodegroups haven't been created properly, you may wish to check CloudFormation console", len(errs))
+			logger.Info("to cleanup resources, run 'eksctl delete nodegroup --region=%s --cluster=%s --name=<name>' for each of the failed nodegroups", cfg.Metadata.Region, cfg.Metadata.Name)
+			for _, err := range errs {
+				if err != nil {
+					logger.Critical("%s\n", err.Error())
+				}
+			}
+			return fmt.Errorf("failed to create nodegroups for cluster %q", cfg.Metadata.Name)
+		}
+
+		for _, ng := range cfg.NodeGroups {
+			if c.options.UpdateAuthConfigMap {
+				// authorise nodes to join
+				if err = authconfigmap.AddNodeGroup(clientSet, ng); err != nil {
+					return err
+				}
+
+				// wait for nodes to join
+				if err = ctl.WaitForNodes(clientSet, ng); err != nil {
+					return err
+				}
+			}
+
+			ShowDevicePluginMessageForNodeGroup(ng, c.options.InstallNeuronDevicePlugin)
+		}
+		logger.Success("created %d nodegroup(s) in cluster %q", len(cfg.NodeGroups), cfg.Metadata.Name)
+
+		for _, ng := range cfg.ManagedNodeGroups {
+			if err := ctl.WaitForNodes(clientSet, ng); err != nil {
+				if cfg.PrivateCluster.Enabled {
+					logger.Info("error waiting for nodes to join the cluster; this command was likely run from outside the cluster's VPC as the API server is not reachable, nodegroup(s) should still be able to join the cluster, underlying error is: %v", err)
+					break
+				} else {
+					return err
+				}
+			}
+		}
+
+		logger.Success("created %d managed nodegroup(s) in cluster %q", len(cfg.ManagedNodeGroups), cfg.Metadata.Name)
+	}
```

Comment:
this is a really long function, can some of the logic get separated out into funcs? The parts where we create a new scope (not sure what the proper term is?) in particular look like good places to move the logic into a descriptive function

Reply:
done